### PR TITLE
SUP-19612: categoryEntry->Add refactor prepare to fix Zoom engine to work with categoryEntry->add vs. entry->setCategories which is deprecated

### PR DESCRIPTION
**Pull request by amiras89**
This change is [Reviewable](https://reviewable.io/reviews/kaltura/server/9233)

### CODE REVIEW

File: api_v3/services/CategoryEntryService.php [hunk 1]
```diff
@@ -28,89 +28,24 @@ public function initService($serviceId, $serviceName, $actionName)
...
-		$partnerId = kCurrentContext::$partner_id ? kCurrentContext::$partner_id : kCurrentContext::$ks_partner_id;
-		$dbCategoryEntry->setPartnerId($partnerId);
-		
-		$kuser = kCurrentContext::getCurrentKsKuser();
-		
-		if ($kuser)
-		{
-			$dbCategoryEntry->setCreatorKuserId($kuser->getId());
-			$dbCategoryEntry->setCreatorPuserId($kuser->getPuserId());
+			if ($ex instanceof kCoreException) {
```
Comment (MosheMaorKaltura) [inline api_v3/services/CategoryEntryService.php]: indentation
if()
{
   do//
}

File: api_v3/services/CategoryEntryService.php [hunk 2]
```diff
@@ -28,89 +28,24 @@ public function initService($serviceId, $serviceName, $actionName)
...
-			$categoryKuser = categoryKuserPeer::retrievePermittedKuserInCategory($categoryEntry->categoryId, kCurrentContext::getCurrentKsKuserId());
-			if(!$categoryKuser ||
-				($categoryKuser->getPermissionLevel() != CategoryKuserPermissionLevel::MANAGER && 
-				$categoryKuser->getPermissionLevel() != CategoryKuserPermissionLevel::MODERATOR))
-				$dbCategoryEntry->setStatus(CategoryEntryStatus::PENDING);
-		}
-		
-		if ($category->getModeration() && 
-		   (kEntitlementUtils::getCategoryModeration() || $this->getPartner()->getEnabledService(KalturaPermissionName::FEATURE_BLOCK_CATEGORY_MODERATION_SELF_APPROVE)))
+		catch (Exception $ex)
```
Comment (MosheMaorKaltura) [inline api_v3/services/CategoryEntryService.php]: No need to check the type of the Exception, in the catch just change the type from Exception to kCoreException,
All Exception that are of different type will be thrown up automatically.

File: alpha/lib/model/categoryEntry.php [hunk 3]
```diff
@@ -219,5 +219,118 @@ public function setCreatorPuserId ($v)
...
+		{
+			throw new kCoreException("Invalid Entry ID: {$entryId}",kCoreException::INVALID_ENTRY_ID, $entryId);
+		}
+
+		$category = categoryPeer::retrieveByPK($categoryId);
+		if(!$category)
+		{
+			throw new kCoreException("Category ID: {$categoryId} not found", kCoreException::CATEGORY_NOT_FOUND, $categoryId);
+		}
+		/*=========================================================*/
```
Comment (MosheMaorKaltura) [inline alpha/lib/model/categoryEntry.php]: remove

File: alpha/lib/model/categoryEntry.php [hunk 4]
```diff
@@ -219,5 +219,118 @@ public function setCreatorPuserId ($v)
...
+		$categoryEntries = categoryEntryPeer::retrieveActiveAndPendingByEntryId($entryId);
+
+		$maxCategoriesPerEntry = $entry->getMaxCategoriesPerEntry();
+
+		if(count($categoryEntries) >= $maxCategoriesPerEntry)
+		{
+			throw new kCoreException("Max categories per entry reached, Allowed: {$maxCategoriesPerEntry}", kCoreException::MAX_CATEGORIES_PER_ENTRY, $maxCategoriesPerEntry);
+		}
+		/*=========================================================*/
```
Comment (MosheMaorKaltura) [inline alpha/lib/model/categoryEntry.php]: remove

**Comment by amiras89** [comment]
> 1. Where is the Zoom part?
> 2. Have you executed BE API tests to verify no regression?

@MosheMaorKaltura 
1. Removed the Zoom part due to major changes coming from Zur
Will wait until he pushes he's code 

2.
a. I made sanity tests: confirmed categoryEntry->add work as expected (writes to DB and return the object to client)
b. Confirmed all exceptions are being thrown when reached.

Should I make other tests?
Thanks

File: alpha/lib/model/categoryEntry.php [hunk 5]
```diff
@@ -219,5 +219,167 @@ public function setCreatorPuserId ($v)
...
+				throw new kCoreException("Cannot assign entry to category", kCoreException::CANNOT_ASSIGN_ENTRY_TO_CATEGORY);
+			}
+		}
+	}
+
+	/**
+	 * @param categoryEntry $categoryEntry
+	 * @throws kCoreException
+	 */
+	protected function checkCategoryEntryNotExist(categoryEntry $categoryEntry = null)
```
Comment (MosheMaorKaltura) [inline alpha/lib/model/categoryEntry.php]: rename - validate.....

File: alpha/lib/model/categoryEntry.php [hunk 6]
```diff
@@ -219,5 +219,167 @@ public function setCreatorPuserId ($v)
...
+			{
+				throw new kCoreException("User '{$currentKsKuserId}' permission level in category Id '{$categoryId}' is 'MEMBER' and is not allowed to add entry to category", kCoreException::CANNOT_ASSIGN_ENTRY_TO_CATEGORY);
+			}
+
+			if(!$categoryKuser->hasPermission(PermissionName::CATEGORY_EDIT) &&
+				!$categoryKuser->hasPermission(PermissionName::CATEGORY_CONTRIBUTE) &&
+				!$entry->isEntitledKuserEdit($currentKsKuserId) &&
+				$entry->getCreatorKuserId() != $currentKsKuserId)
+			{
+				throw new kCoreException("Cannot assign entry to category", kCoreException::CANNOT_ASSIGN_ENTRY_TO_CATEGORY);
```
Comment (MosheMaorKaltura) [inline alpha/lib/model/categoryEntry.php]: Use '

File: alpha/lib/model/categoryEntry.php [hunk 7]
```diff
@@ -219,5 +219,167 @@ public function setCreatorPuserId ($v)
...
+			$this->setCreatorPuserId($kuser->getPuserId());
+		}
+	}
+
+	/**
+	 * @param $entryId
+	 * @return entry
+	 * @throws kCoreException
+	 */
+	protected function validateAndReturnEntry($entryId)
```
Comment (MosheMaorKaltura) [inline alpha/lib/model/categoryEntry.php]: rename retrieve...

File: alpha/lib/model/categoryEntry.php [hunk 8]
```diff
@@ -219,5 +219,167 @@ public function setCreatorPuserId ($v)
...
+
+		$this->validateMaxCategoriesPerEntry($entry);
+		$currentKsKuserId = kCurrentContext::getCurrentKsKuserId();
+		$this->validateKuserEntitledToAssignEntryToCategory($category, $entry, $currentKsKuserId);
+		$categoryEntryExists = categoryEntryPeer::retrieveByCategoryIdAndEntryId($categoryId, $entryId);
+		$this->checkCategoryEntryNotExist($categoryEntryExists);
+
+		if($categoryEntryExists)
+		{
+			$this->copyFrom($categoryEntryExists);
```
Comment (MosheMaorKaltura) [inline alpha/lib/model/categoryEntry.php]: use categoryEntryExists->copyInto($this)

**Pull request closed without merging**